src/passes/OptimizeInstructions.cpp

```diff
@@ -2874,20 +2874,47 @@ struct OptimizeInstructions
           // TODO: consider the case with more than one child.
           ChildIterator ifTrueChildren(curr->ifTrue);
           if (ifTrueChildren.children.size() == 1) {
-            ChildIterator ifFalseChildren(curr->ifFalse);
             // ifTrue and ifFalse's children will become the direct children of
-            // curr, and so there must be an LUB for curr to have a proper type.
+            // curr, and so there must be an LUB for curr to have a proper new
+            // type after the transformation.
+            //
             // An example where that does not happen is this:
             //
             //  (if
             //    (condition)
             //    (drop (i32.const 1))
             //    (drop (f64.const 2.0))
             //  )
+            ChildIterator ifFalseChildren(curr->ifFalse);
             auto* ifTrueChild = *ifTrueChildren.begin();
             auto* ifFalseChild = *ifFalseChildren.begin();
-            bool validTypes =
-              Type::hasLeastUpperBound(ifTrueChild->type, ifFalseChild->type);
+            auto newCurrType =
+              Type::getLeastUpperBound(ifTrueChild->type, ifFalseChild->type);
+            bool validTypes = newCurrType != Type::none;
```

Comment:
Why switch to `getLeastUpperBound` here? `hasLeastUpperBound` is more efficient when the result is not used.

Reply:
Oh, I think it was clearer to have `newCurrType` to match `newOuterType`. But yeah, not worth it I guess...